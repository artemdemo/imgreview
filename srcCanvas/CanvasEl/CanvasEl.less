.canvas-container {
  overflow: hidden;
  margin: 0 auto;
  background: rgba(0, 0, 0, 0.03);
}

.canvas-scroll {
  width: 100%;
  overflow: auto;
}

.canvas-el {
  outline: none;
}
