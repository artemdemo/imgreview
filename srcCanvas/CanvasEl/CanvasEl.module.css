.CanvasEl {
  outline: none;
  width: 100%;
  height: 100%;
}
