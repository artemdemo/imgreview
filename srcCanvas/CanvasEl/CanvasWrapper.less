.CanvasWrapper {
  overflow: hidden;
  margin: 0 auto;
  height: 100%;
  background: rgba(0, 0, 0, 0.03);
}
