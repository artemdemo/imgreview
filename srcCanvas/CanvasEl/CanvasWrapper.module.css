.CanvasWrapper {
  overflow: hidden;
  height: 100%;
  width: 100%;
  background: rgba(0, 0, 0, 0.03);
  position: absolute;
}
