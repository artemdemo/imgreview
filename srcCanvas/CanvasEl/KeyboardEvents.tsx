import React, { useEffect, useRef } from 'react';
import { GlobalHotKeys } from 'react-hotkeys';
import _ from 'lodash';
import canvasStore from '../store';
import {
  blurShapes,
  deleteActiveShapes,
  setCursor,
} from '../model/shapes/shapesActions';
import { ECursorTypes } from '../model/shapes/shapesModelTypes';
import { TCanvasState } from '../reducers';
import * as clipboard from '../services/clipboard';
import Shape from '../canvasShapes/Shape/Shape';
import { cloneAndConnectShape } from '../addShape';
import { TOneOfShapeTypes } from '../model/shapes/shapesReducer';
import { setRatioShift, setStageDraggable } from '../model/stage/stageActions';

const keyMap = {
  onDelete: ['backspace', 'delete', 'del'],
  onCopy: ['ctrl+c', 'command+c'],
  onPaste: ['ctrl+v', 'command+v'],
  onActivateDrag: {
    sequence: 'space',
    action: 'keydown',
  },
  onDisableDrag: {
    sequence: 'space',
    action: 'keyup',
  },
  onActivateRatio: {
    sequence: 'shift',
    action: 'keydown',
  },
  onDisableRatio: {
    sequence: 'shift',
    action: 'keyup',
  },
};

export const KeyboardEvents: React.FC = () => {
  const copiedShapes = useRef<TOneOfShapeTypes[]>([]);

  useEffect(() => {
    const onTabFocus = _.throttle(() => {
      onDisableDrag();
      onDisableRatio();
    }, 50);
    window.addEventListener('focus', onTabFocus, { capture: true });
    return () => {
      window.removeEventListener('focus', onTabFocus);
    };
  }, []);

  const onDelete = () => {
    canvasStore.dispatch(deleteActiveShapes());
    // In case users cursor is on the shape that is being deleted.
    // I need to remove cursor styling from the parent.
    canvasStore.dispatch(setCursor(ECursorTypes.AUTO));
  };

  const onCopy = () => {
    const { shapes } = canvasStore.getState();

    // Here I'm coping a dummy text into the clipboard.
    // This is workaround for case when user has image in the clipboard.
    // In this scenario after copying a shape and pasting it - image will appear and override everything.
    // ToDo: Copy image of the shape into the clipboard.
    //       This approach will be much better experience.
    clipboard.copyToClipboard('[Shape]');

    copiedShapes.current = shapes.list.reduce((acc: any[], shape) => {
      if (shape.isSelected()) {
        return [
          ...acc,
          // I need to clone here,
          // so copied shape will keep exact coordinates of the moment of copying
          shape.clone(),
        ];
      }
      return acc;
    }, []);
  };

  const onPaste = () => {
    canvasStore.dispatch(blurShapes());
    copiedShapes.current.forEach((shape: Shape) => {
      cloneAndConnectShape(shape);
    });
    copiedShapes.current = [];
  };

  const onActivateDrag = () => {
    const { shapes } = canvasStore.getState();
    canvasStore.dispatch(setStageDraggable(true));
    // User should be able to drag stage holding by any part of the canvas.
    // Even if this "part" is another shape.
    // Therefore, I'm while "space" is clicked all shapes should be not draggable.
    shapes.list.forEach((item) => {
      item.draggable(false);
    });
    canvasStore.dispatch(setCursor(ECursorTypes.GRAB));
  };

  const onDisableDrag = () => {
    const { shapes } = canvasStore.getState();
    canvasStore.dispatch(setStageDraggable(false));
    shapes.list.forEach((item) => {
      item.draggable(true);
    });
    canvasStore.dispatch(setCursor(ECursorTypes.AUTO));
  };

  const onActivateRatio = () => {
    canvasStore.dispatch(setRatioShift(true));
  };

  const onDisableRatio = () => {
    canvasStore.dispatch(setRatioShift(false));
  };

  return (
    <GlobalHotKeys
      // @ts-ignore
      keyMap={keyMap}
      handlers={{
        onDelete,
        onCopy,
        onPaste,
        onActivateDrag,
        onDisableDrag,
        onActivateRatio,
        onDisableRatio,
      }}
    />
  );
};
