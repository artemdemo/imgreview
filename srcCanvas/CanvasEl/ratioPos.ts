import { TOneOfShapeTypes } from '../model/shapes/shapesReducer';
import { TPos } from 'konva';
import EShapeTypes from '../canvasShapes/Shape/shapeTypes';

type ApplyInitDrawOptions = {
  shape: TOneOfShapeTypes;
  startPos: TPos;
  currentPos: TPos;
  ratioShiftIsActive: boolean;
};
export const applyInitDraw = (options: ApplyInitDrawOptions): void => {
  const { shape, startPos, currentPos, ratioShiftIsActive } = options;
  if (ratioShiftIsActive) {
    if (shape.type !== EShapeTypes.ARROW && shape.type !== EShapeTypes.TEXT) {
      const horizontalDiff = currentPos.x - startPos.x;
      const verticalDiff = currentPos.y - startPos.y;
      const ratioWidth = Math.min(
        Math.abs(horizontalDiff),
        Math.abs(verticalDiff),
      );
      shape.initDraw(startPos, {
        x: startPos.x + Math.sign(horizontalDiff) * ratioWidth,
        y: startPos.y + Math.sign(verticalDiff) * ratioWidth,
      });
      return;
    }
  }
  shape.initDraw(startPos, currentPos);
};
