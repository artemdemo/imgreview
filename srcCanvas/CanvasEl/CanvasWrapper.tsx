import React, { useEffect, useRef } from 'react';
import _ from 'lodash';
import canvasStore from '../store';
import './CanvasWrapper.less';
import { setStageSize } from '../model/stage/stageActions';

export const CanvasWrapper: React.FC = (props) => {
  const { children } = props;

  const handleWindowSize = useRef(
    _.throttle(() => {
      canvasStore.dispatch(
        setStageSize({
          width: window.innerWidth,
          height: window.innerHeight,
        })
      );
    }, 80)
  );

  useEffect(() => {
    window.addEventListener('resize', handleWindowSize.current);
    return () => {
      window.removeEventListener('resize', handleWindowSize.current);
    };
  }, []);

  return <div className="CanvasWrapper">{children}</div>;
};
