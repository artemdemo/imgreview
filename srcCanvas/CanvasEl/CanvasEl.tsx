import React from 'react';
import { TPos } from 'konva';
import { setAddingShape, drawLayers } from '../model/shapes/shapesActions';
import { connectShape } from '../addShape';
import { TCanvasState } from '../reducers';
import canvasStore from '../store';
import { setStage } from '../model/stage/stageActions';
import { setSaveStage } from '../model/saveCanvas/saveCanvasActions';
import { SHAPES_LAYER_CLS } from '../model/shapes/shapesConst';
import { KeyboardEvents } from './KeyboardEvents';
import { CanvasWrapper } from './CanvasWrapper';
import Stage from './Stage';
import { SaveStage } from './SaveStage';
import '../events/events';
import './CanvasEl.less';
import { SaveCanvasEl } from '../SaveCanvasEl/SaveCanvasEl';

export const getShapesLayerEl = (): HTMLCanvasElement => {
  const shapesLayerEl: HTMLCanvasElement | null = document.querySelector(
    `.${SHAPES_LAYER_CLS}`
  );
  if (shapesLayerEl) {
    return shapesLayerEl;
  }
  throw new Error(`Shapes layer is not found`);
};

type Props = {};

type State = {
  mouseIsDown: boolean;
  mouseStartPos: TPos;
};

/**
 * CanvasEl will be used inside the main app.
 * Therefore, I can't use `connect()` here, since the context will be of the main app and not of the canvas.
 */
class CanvasEl extends React.PureComponent<Props, State> {
  readonly canvasRef = React.createRef<HTMLDivElement>();

  state = {
    mouseIsDown: false,
    mouseStartPos: { x: 0, y: 0 },
  };

  componentDidMount() {
    if (this.canvasRef.current) {
      const stage = new Stage(this.canvasRef.current);
      stage.on('mousedown', this.handleStageOnMouseDown);
      stage.on('mouseup', this.handleStageOnMouseUp);
      stage.on('mousemove', this.handleStageOnMouseMove);
      canvasStore.dispatch(setStage(stage));
      this.canvasRef.current.tabIndex = 1;
    }
  }

  private handleStageOnMouseDown = (e: any) => {
    const { shapes, stage } = canvasStore.getState() as TCanvasState;
    if (shapes.addingShapeRef && stage.instance) {
      const absPos = stage.instance.absolutePosition()!;
      const { layerX, layerY } = e.evt;
      const mouseStartPos = {
        x: layerX - absPos.x,
        y: layerY - absPos.y,
      };
      this.setState({
        mouseIsDown: true,
        mouseStartPos,
      });
      connectShape(shapes.addingShapeRef);
      shapes.addingShapeRef.initDraw(mouseStartPos, mouseStartPos);
    }
  };

  private handleStageOnMouseUp = () => {
    this.setState({ mouseIsDown: false });
    const { shapes } = canvasStore.getState() as TCanvasState;
    if (shapes.addingShapeRef) {
      shapes.addingShapeRef.focus();
    }
    canvasStore.dispatch(setAddingShape(null));
    // I need to redraw shapes in order to focus to take effect.
    canvasStore.dispatch(drawLayers());
  };

  private handleStageOnMouseMove = (e: any) => {
    const { shapes, stage } = canvasStore.getState() as TCanvasState;
    if (this.state.mouseIsDown && shapes.addingShapeRef && stage.instance) {
      const absPos = stage.instance.absolutePosition()!;
      const { layerX, layerY } = e.evt;
      shapes.addingShapeRef.initDraw(this.state.mouseStartPos, {
        x: layerX - absPos.x,
        y: layerY - absPos.y,
      });
    }
  };

  render() {
    return (
      <>
        <CanvasWrapper>
          <KeyboardEvents />
          <div ref={this.canvasRef} className="canvas-el" />
        </CanvasWrapper>
        <SaveCanvasEl />
      </>
    );
  }
}

// not using `connect()` see reason in the comment before class definition
export default CanvasEl;
