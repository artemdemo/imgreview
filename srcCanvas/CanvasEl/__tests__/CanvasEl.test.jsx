import React from 'react';
import { mount } from 'enzyme/build';
import CanvasEl from '../CanvasEl';

jest.mock('konva');
jest.mock('react-redux');

describe('CanvasEl', () => {
    jest.clearAllMocks();
    const state = {
        canvas: {
            cursor: 'auto',
        },
    };
    const konvaMock = require('konva/konva');
    const reactReduxMock = require('react-redux');

    it('should create stage', () => {
        const wrapper = mount(
            <CanvasEl
                canvas={state.canvas}
            />
        );
        const { canvasRef } = wrapper.instance();
        expect(konvaMock.Stage).toBeCalledWith({
            container: canvasRef.current,
        });
        expect(reactReduxMock.__getLastMaps().mapStateToProps(state)).toEqual({
            canvas: state.canvas,
        });
    });

    // it('should blur shapes on stage click', () => {
    //     konvaMock.__clearStageCallbacks();
    //     const setStageMock = jest.fn();
    //     const blurShapesMock = jest.fn();
    //     mount(
    //         <CanvasEl
    //             setStage={setStageMock}
    //             blurShapes={blurShapesMock}
    //             canvas={state.canvas}
    //         />
    //     );
    //     konvaMock.__callStage('click');
    //     expect(blurShapesMock).toBeCalled();
    // });
});
