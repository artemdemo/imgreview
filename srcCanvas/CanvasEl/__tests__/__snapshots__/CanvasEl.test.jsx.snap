// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasEl default render 1`] = `
<div
  className="canvas-scroll"
>
  <div
    data-mock="GlobalHotKeys"
  >
    {
  "keyMap": {
    "delete": [
      "backspace",
      "delete",
      "del"
    ],
    "copy": [
      "ctrl+c",
      "command+c"
    ],
    "paste": [
      "ctrl+v",
      "command+v"
    ]
  },
  "handlers": {}
}
  </div>
  <div
    className="canvas-container"
    style={
      Object {
        "height": 0,
        "width": 0,
      }
    }
  >
    <div
      className="canvas-el"
      onClick={[Function]}
      style={
        Object {
          "cursor": "auto",
        }
      }
    />
  </div>
</div>
`;
